File: src/main/java/top/chenray/antilitematica/guard/PlacementGuard.java
package top.chenray.antilitematica.guard;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import org.bukkit.Bukkit;
import org.bukkit.FluidCollisionMode;
import org.bukkit.GameMode;
import org.bukkit.Location;
import org.bukkit.block.Block;
import org.bukkit.entity.Player;
import org.bukkit.event.Cancellable;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockMultiPlaceEvent;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.event.player.PlayerChangedWorldEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.util.RayTraceResult;
import top.chenray.antilitematica.AntiLitematicaPlugin;
import top.chenray.antilitematica.api.event.DetectionEvent;
import top.chenray.antilitematica.api.event.PunishmentEvent;
import top.chenray.antilitematica.config.Settings;
import top.chenray.antilitematica.util.Msg;
import top.chenray.antilitematica.util.SchedulerUtil;
import top.chenray.antilitematica.util.TokenBucket;
import top.chenray.antilitematica.util.ViolationWindow;

public final class PlacementGuard implements Listener {
   private final AntiLitematicaPlugin plugin;
   private final Settings settings;
   /** Unified per-player tracking: player UUID -> tracker. Single map, single cleanup point. */
   private final Map<UUID, PlayerTracker> trackers = new ConcurrentHashMap<>();
   /** Cached set of staff UUIDs with antilitematica.notify permission. */
   private final Set<UUID> staffNotify = ConcurrentHashMap.newKeySet();

   // Cached config values, refreshed on start/reload
   private volatile int cachedMaxBlocksPerSecond = 14;
   private volatile int cachedKickAt = 8;
   private volatile boolean apEnabled;
   private volatile boolean applyToCreative;
   private volatile boolean enforceRaytrace;
   private volatile boolean detectConsecutiveSameType;
   private volatile boolean detectNoLookChange;
   private volatile long consecutiveWindowMs = 3000L; // now configurable

   public PlacementGuard(AntiLitematicaPlugin plugin, Settings settings) {
      this.plugin = plugin;
      this.settings = settings;
   }

   public void start() {
      refreshConfigCache();
      if (this.apEnabled) {
         Bukkit.getPluginManager().registerEvents(this, this.plugin);
         // Pre-populate staff notify cache
         for (Player online : Bukkit.getOnlinePlayers()) {
            if (online.hasPermission("antilitematica.notify")) {
               staffNotify.add(online.getUniqueId());
            }
         }
      }
   }

   /** Reload cached config values to avoid repeated getter chain calls at runtime. */
   public void refreshConfigCache() {
      Settings.AntiPrinter ap = this.settings.antiPrinter();
      this.apEnabled = ap.enabled();
      this.applyToCreative = ap.applyToCreative();
      this.enforceRaytrace = ap.enforceRaytrace();
      this.detectConsecutiveSameType = ap.detectConsecutiveSameType();
      this.detectNoLookChange = ap.detectNoLookChange();
      this.cachedMaxBlocksPerSecond = Math.max(1, plugin.getDynamicThresholdManager().adjustInt(ap.maxBlocksPerSecond()));
      this.cachedKickAt = plugin.getDynamicThresholdManager().adjustInt(ap.violations().kickAt());
      this.consecutiveWindowMs = Math.max(1000L, this.settings.antiPrinter().violations().windowMs() / 2); // use half the violation window
   }

   public void shutdown() {
      HandlerList.unregisterAll(this);
      this.trackers.clear();
      this.staffNotify.clear();
   }

   /** Get world key for map lookups. */
   @EventHandler(
      priority = EventPriority.HIGHEST,
      ignoreCancelled = true
   )
   public void onPlace(BlockPlaceEvent event) {
      if (event instanceof BlockMultiPlaceEvent) return;
      if (!apEnabled) return;
      Player p = event.getPlayer();
      if (shouldBypass(p)) return;
      // Check per-world anti-printer config
      if (!this.settings.isAntiPrinterEnabledForWorld(p.getWorld().getName())) return;
      if (!applyToCreative && p.getGameMode() == GameMode.CREATIVE) return;

      if (cachedMaxBlocksPerSecond > 0 && !this.consume(p, 1)) {
         this.deny(event, p, "rate");
         return;
      }
      if (this.enforceRaytrace && !this.rayTraceMatches(p, event.getBlockPlaced(), event.getBlockAgainst())) {
         // Tweakeroo flexi placement modifies hit vectors — skip raytrace in compat mode
         if (!this.plugin.getMasaCompat().shouldSkipRaytrace(p)) {
            this.deny(event, p, "raytrace");
            return;
         }
      }
      if (this.detectConsecutiveSameType && this.checkConsecutiveSameType(p, event.getBlockPlaced())) {
         this.deny(event, p, "consecutive_same");
         return;
      }
      if (this.detectNoLookChange && this.checkNoLookChange(p)) {
         this.deny(event, p, "no_look_change");
      }
   }

   @EventHandler(
      priority = EventPriority.HIGHEST,
      ignoreCancelled = true
   )
   public void onMultiPlace(BlockMultiPlaceEvent event) {
      if (!apEnabled) return;
      Player p = event.getPlayer();
      if (shouldBypass(p)) return;
      // Check per-world anti-printer config
      if (!this.settings.isAntiPrinterEnabledForWorld(p.getWorld().getName())) return;
      if (!applyToCreative && p.getGameMode() == GameMode.CREATIVE) return;

      int count = Math.max(1, event.getReplacedBlockStates().size());
      if (cachedMaxBlocksPerSecond > 0 && !this.consume(p, count)) {
         this.deny(event, p, "rate");
         return;
      }
      if (this.enforceRaytrace && !this.rayTraceMatches(p, event.getBlockPlaced(), event.getBlockAgainst())) {
         if (!this.plugin.getMasaCompat().shouldSkipRaytrace(p)) {
            this.deny(event, p, "raytrace");
            return;
         }
      }
      if (this.detectConsecutiveSameType && this.checkConsecutiveSameType(p, event.getBlockPlaced())) {
         this.deny(event, p, "consecutive_same");
         return;
      }
      if (this.detectNoLookChange && this.checkNoLookChange(p)) {
         this.deny(event, p, "no_look_change");
      }
   }

   @EventHandler
   public void onJoin(PlayerJoinEvent event) {
      Player p = event.getPlayer();
      if (p.hasPermission("antilitematica.notify")) {
         staffNotify.add(p.getUniqueId());
      }
   }

   @EventHandler
   public void onQuit(PlayerQuitEvent event) {
      UUID id = event.getPlayer().getUniqueId();
      this.trackers.remove(id);
      this.staffNotify.remove(id);
   }

   @EventHandler
   public void onWorldChange(PlayerChangedWorldEvent event) {
      // Reset tracker state for the new world (fresh start in different world)
      Player p = event.getPlayer();
      PlayerTracker tracker = this.trackers.get(p.getUniqueId());
      if (tracker != null) tracker.resetState();
   }

   private boolean consume(Player p, int blocks) {
      PlayerTracker tracker = this.trackers.computeIfAbsent(p.getUniqueId(), k -> new PlayerTracker(cachedMaxBlocksPerSecond));
      if (tracker.bucket == null) {
         tracker.bucket = TokenBucket.perSecond(cachedMaxBlocksPerSecond, Math.max(cachedMaxBlocksPerSecond * 2L, 10L));
      }
      return tracker.bucket.tryConsume(blocks);
   }

   private void deny(Cancellable event, Player p, String type) {
      // Check world whitelist
      if (this.settings.worldWhitelist() != null
            && this.settings.worldWhitelist().isWorldExempt(p.getWorld().getName())) {
         return;
      }
      event.setCancelled(true);
      PlayerTracker tracker = this.trackers.computeIfAbsent(p.getUniqueId(), k -> new PlayerTracker(cachedMaxBlocksPerSecond));
      if (tracker.violations == null) {
         tracker.violations = new ViolationWindow(this.settings.antiPrinter().violations().windowMs());
      }
      int n = tracker.violations.addViolation();

      if (this.settings.integration().enabled()) {
         Settings.Integration integ = this.settings.integration();
         this.plugin.getIntegrationManager().flag(p, integ.checkPrefix() + ":printer_" + type,
               integ.violationLevel(), "printer detection: " + type);
      }

      if (n == 1 || n % 3 == 0) {
         p.sendMessage(Msg.color(Msg.prefix(this.settings) + this.settings.messages().blockedPlace()));
      }

      // ---- Fire DetectionEvent ----
      DetectionEvent detectionEvent = new DetectionEvent(p, "printer", type, DetectionEvent.DetectionType.PRINTER);
      Bukkit.getPluginManager().callEvent(detectionEvent);

      if (n >= cachedKickAt) {
         if (p.isOnline() && !detectionEvent.isCancelled()) {
            this.plugin.getLogger().info("Kicking " + p.getName() + " due to repeated blocked placements (" + type + "), violations=" + n);
            final String kickMsgStr = Msg.color(Msg.prefix(this.settings) + this.settings.messages().kick());
            SchedulerUtil.runForPlayer(this.plugin, p, () -> {
               if (p.isOnline() && !shouldBypass(p)) {
                  p.kickPlayer(kickMsgStr);
               }
            });
            // ---- Fire PunishmentEvent ----
            SchedulerUtil.runGlobal(this.plugin, () -> {
               try {
                  PunishmentEvent pe = new PunishmentEvent(p, "printer", type,
                        PunishmentEvent.PunishmentAction.KICK, n, "printer");
                  Bukkit.getPluginManager().callEvent(pe);
               } catch (Exception ignored) {}
            });
         }
      } else {
         this.notifyStaff(p, type, n);
      }
   }

   private void notifyStaff(Player p, String type, int violations) {
      String msg = Msg.color(Msg.prefix(this.settings) + "&e" + p.getName()
            + " &7blocked placement (&f" + type + "&7), vio=&f" + violations);
      for (UUID uid : staffNotify) {
         Player staff = Bukkit.getPlayer(uid);
         if (staff != null && staff.isOnline()) {
            staff.sendMessage(msg);
         }
      }
   }

   private boolean rayTraceMatches(Player p, Block placed, Block against) {
      double reach = (p.getGameMode() == GameMode.CREATIVE
            ? this.settings.antiPrinter().reachCreative()
            : this.settings.antiPrinter().reachSurvival())
            + this.settings.antiPrinter().extraReachAllowance();
      RayTraceResult r = p.rayTraceBlocks(reach, FluidCollisionMode.NEVER);
      if (r == null) return false;
      Block hit = r.getHitBlock();
      return hit != null && (sameBlock(hit, against) || sameBlock(hit, placed));
   }

   private boolean checkConsecutiveSameType(Player p, Block placed) {
      UUID id = p.getUniqueId();
      PlayerTracker tracker = this.trackers.computeIfAbsent(id, k -> new PlayerTracker(cachedMaxBlocksPerSecond));
      if (tracker.recentPlacements == null) {
         tracker.recentPlacements = new ArrayDeque<>();
      }
      Deque<PlacementSnapshot> deque = tracker.recentPlacements;
      long now = System.currentTimeMillis();
      String type = placed.getType().name();
      deque.addLast(new PlacementSnapshot(now, type));
      // Remove entries older than configured window
      while (!deque.isEmpty() && now - deque.peekFirst().time > consecutiveWindowMs) {
         deque.pollFirst();
      }
      int threshold = Math.max(3, plugin.getDynamicThresholdManager().adjustInt(
            this.settings.antiPrinter().consecutiveSameTypeThreshold()));
      // Require at least some placements to avoid premature flagging
      if (deque.size() < threshold) return false;

      // Use sliding ratio: flag if the dominant type accounts for >80% of placements
      // This is smarter than requiring ALL to be same type (human builders mix blocks)
      int sameCount = 0;
      for (PlacementSnapshot snap : deque) {
         if (snap.blockType.equals(type)) sameCount++;
      }
      double ratio = (double) sameCount / deque.size();
      if (ratio >= 0.8) {
         deque.clear();
         return true;
      }
      return false;
   }

   private boolean checkNoLookChange(Player p) {
      UUID id = p.getUniqueId();
      PlayerTracker tracker = this.trackers.computeIfAbsent(id, k -> new PlayerTracker(cachedMaxBlocksPerSecond));
      Location loc = p.getLocation();
      float yaw = loc.getYaw();
      float pitch = loc.getPitch();
      Float lastY = tracker.lastYaw;
      Float lastP = tracker.lastPitch;
      tracker.lastYaw = yaw;
      tracker.lastPitch = pitch;
      if (lastY != null && lastP != null) {
         float yawDiff = Math.abs(yaw - lastY);
         float pitchDiff = Math.abs(pitch - lastP);
         // Normalize yaw diff to 0..180
         if (yawDiff > 180.0f) yawDiff = 360.0f - yawDiff;
         // Use different tolerances: yaw is stricter (0.05), pitch allows a bit more (0.1)
         // because vertical mouse movement is more common during building
         if (yawDiff < 0.05f && pitchDiff < 0.1f) {
            tracker.noLookCount++;
            if (tracker.noLookCount >= Math.max(2, plugin.getDynamicThresholdManager().adjustInt(
                  this.settings.antiPrinter().noLookChangeThreshold()))) {
               tracker.noLookCount = 0;
               return true;
            }
         } else {
            tracker.noLookCount = 0;
         }
      }
      return false;
   }

   private static boolean sameBlock(Block a, Block b) {
      if (a != null && b != null) {
         if (a.getWorld() != b.getWorld()) {
            return false;
         } else {
            return a.getX() == b.getX() && a.getY() == b.getY() && a.getZ() == b.getZ();
         }
      } else {
         return false;
      }
   }

   private static boolean shouldBypass(Player p) {
      return p.hasPermission("antilitematica.bypass")
            || top.chenray.antilitematica.util.BedrockPlayerDetector.isBedrockPlayer(p);
   }

   /**
    * Unified per-player tracking data. Replaces 6 separate maps with 1.
    * Player leaves → one remove() call cleans everything.
    * Player changes world → resetState() resets counters for fresh start.
    */
   private static final class PlayerTracker {
      TokenBucket bucket;
      ViolationWindow violations;
      Deque<PlacementSnapshot> recentPlacements;
      float lastYaw = Float.NaN;
      float lastPitch = Float.NaN;
      int noLookCount;

      PlayerTracker(int maxBlocksPerSecond) {
         this.bucket = TokenBucket.perSecond(maxBlocksPerSecond, Math.max(maxBlocksPerSecond * 2L, 10L));
         this.violations = null; // Created lazily in deny()
      }

      void resetState() {
         this.bucket = null; // Recreated on next consume()
         this.violations = null;
         if (this.recentPlacements != null) this.recentPlacements.clear();
         this.lastYaw = Float.NaN;
         this.lastPitch = Float.NaN;
         this.noLookCount = 0;
      }
   }

   private static final class PlacementSnapshot {
      final long time;
      final String blockType;
      PlacementSnapshot(long time, String blockType) {
         this.time = time;
         this.blockType = blockType;
      }
   }
}

File: src/main/kotlin/icu/epochcraft/antilitematica/guard/PlacementGuard.kt
package icu.epochcraft.antilitematica.guard

import icu.epochcraft.antilitematica.AntiLitematica
import icu.epochcraft.antilitematica.event.DetectionType
import icu.epochcraft.antilitematica.util.TokenBucket
import org.bukkit.FluidCollisionMode
import org.bukkit.GameMode
import org.bukkit.Location
import org.bukkit.block.Block
import org.bukkit.entity.Player
import org.bukkit.event.Cancellable
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockMultiPlaceEvent
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.event.player.PlayerChangedWorldEvent
import org.bukkit.event.player.PlayerQuitEvent
import java.util.ArrayDeque
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * 防 Printer（Litematica 打印机自动放置检测）。
 *
 * 检测手段（移植自旧版 PlacementGuard）：
 *   1. 令牌桶限速（max-blocks-per-second）——打印机放置速度远超人类
 *   2. 射线校验（enforce-raytrace）——Tweakeroo 灵活放置会偏离真实射线
 *   3. 连续同类型方块（比例 >= 80%）
 *   4. 视角不变检测（yaw/pitch 无变化连续放置）
 *
 * 与旧版的关键差异：本实现只负责"判定 + 取消事件"，惩罚统一走
 * [icu.epochcraft.antilitematica.detection.DetectionBus]（渐进惩罚或基础动作），
 * 不再有独立的惩罚路径（修复旧版 B5）。
 *
 * @author 阿清
 */
class PlacementGuard(private val plugin: AntiLitematica) : Listener {

    private data class PlayerTracker(
        var bucket: TokenBucket? = null,
        var recentPlacements: ArrayDeque<PlacementSnapshot>? = null,
        var lastYaw: Float? = null,
        var lastPitch: Float? = null,
    )

    private data class PlacementSnapshot(val time: Long, val blockType: String)

    private val trackers = ConcurrentHashMap<UUID, PlayerTracker>()

    // ---- 缓存配置（reload 时刷新） ----
    private var enabled = false
    private var maxBlocksPerSecond = 14
    private var applyToCreative = false
    private var enforceRaytrace = true
    private var detectConsecutive = true
    private var consecutiveThreshold = 8
    private var consecutiveWindowMs = 3000L
    private var detectNoLook = true
    private var reachSurvival = 4.5
    private var reachCreative = 5.0
    private var extraReachAllowance = 0.5

    /** reload 配置缓存 */
    fun reload() {
        val ap = plugin.configHolder.antiPrinter
        enabled = ap.enabled
        maxBlocksPerSecond = ap.maxBlocksPerSecond.coerceAtLeast(1)
        applyToCreative = ap.applyToCreative
        enforceRaytrace = ap.enforceRaytrace
        detectConsecutive = ap.detectConsecutiveSameType
        consecutiveThreshold = ap.consecutiveSameTypeThreshold.coerceAtLeast(3)
        consecutiveWindowMs = ap.consecutiveWindowMs.coerceAtLeast(1000)
        detectNoLook = ap.detectNoLookChange
        reachSurvival = ap.reachSurvival
        reachCreative = ap.reachCreative
        extraReachAllowance = ap.extraReachAllowance
        if (!enabled) trackers.clear()
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onPlace(event: BlockPlaceEvent) {
        if (event is BlockMultiPlaceEvent) return
        handlePlace(event, event.player, event.blockPlaced, event.blockAgainst, 1)
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onMultiPlace(event: BlockMultiPlaceEvent) {
        val count = event.replacedBlockStates.size.coerceAtLeast(1)
        handlePlace(event, event.player, event.blockPlaced, event.blockAgainst, count)
    }

    private fun handlePlace(event: Cancellable, p: Player, placed: Block, against: Block, count: Int) {
        if (!enabled) return
        if (p.hasPermission("antilitematica.bypass")) return
        if (!applyToCreative && p.gameMode == GameMode.CREATIVE) return

        // 1. 限速
        if (maxBlocksPerSecond > 0 && !consume(p, count)) {
            deny(event, p, "rate")
            return
        }
        // 2. 射线校验
        if (enforceRaytrace && !rayTraceMatches(p, placed, against)) {
            deny(event, p, "raytrace")
            return
        }
        // 3. 连续同类型
        if (detectConsecutive && checkConsecutiveSameType(p, placed)) {
            deny(event, p, "consecutive_same")
            return
        }
        // 4. 视角不变
        if (detectNoLook && checkNoLookChange(p)) {
            deny(event, p, "no_look_change")
        }
    }

    /** 判定命中：取消事件并发射到检测总线（统一惩罚管线） */
    private fun deny(event: Cancellable, p: Player, type: String) {
        event.isCancelled = true
        plugin.detectionBus.emit(
            p,
            "printer:$type",
            plugin.configHolder.lang("detection.ban-reason").replace("{channel}", "printer:$type"),
            DetectionType.PRINTER,
        )
    }

    // ---------------- 检测逻辑 ----------------

    private fun consume(p: Player, blocks: Int): Boolean {
        val tracker = trackers.computeIfAbsent(p.uniqueId) { PlayerTracker() }
        if (tracker.bucket == null) {
            tracker.bucket = TokenBucket.perSecond(maxBlocksPerSecond, maxBlocksPerSecond * 2L.toInt().coerceAtLeast(10))
        }
        return tracker.bucket!!.tryConsume(blocks)
    }

    private fun rayTraceMatches(p: Player, placed: Block, against: Block): Boolean {
        val reach = (if (p.gameMode == GameMode.CREATIVE) reachCreative else reachSurvival) + extraReachAllowance
        val result = p.rayTraceBlocks(reach, FluidCollisionMode.NEVER) ?: return false
        val hit = result.hitBlock ?: return false
        return hit == against || hit == placed
    }

    private fun checkConsecutiveSameType(p: Player, placed: Block): Boolean {
        val tracker = trackers.computeIfAbsent(p.uniqueId) { PlayerTracker() }
        val deque = tracker.recentPlacements ?: ArrayDeque<PlacementSnapshot>().also { tracker.recentPlacements = it }
        val now = System.currentTimeMillis()
        val type = placed.type.name

        deque.addLast(PlacementSnapshot(now, type))
        while (deque.isNotEmpty() && now - deque.peekFirst().time > consecutiveWindowMs) {
            deque.pollFirst()
        }
        if (deque.size < consecutiveThreshold) return false

        // 窗口内主导类型占比 >= 80% 判定为打印机（人类建造会混合方块）
        val sameCount = deque.count { it.blockType == type }
        if (sameCount.toDouble() / deque.size >= 0.8) {
            deque.clear()
            return true
        }
        return false
    }

    private fun checkNoLookChange(p: Player): Boolean {
        val tracker = trackers.computeIfAbsent(p.uniqueId) { PlayerTracker() }
        val loc: Location = p.location
        val yaw = loc.yaw
        val pitch = loc.pitch

        val lastYaw = tracker.lastYaw
        val lastPitch = tracker.lastPitch
        tracker.lastYaw = yaw
        tracker.lastPitch = pitch

        // 视角完全没变（差值 < 0.05°）连续放置 = 自动化特征
        return lastYaw != null && lastPitch != null &&
            kotlin.math.abs(yaw - lastYaw) < 0.05f && kotlin.math.abs(pitch - lastPitch) < 0.05f
    }

    // ---------------- 状态清理 ----------------

    @EventHandler
    fun onQuit(event: PlayerQuitEvent) {
        trackers.remove(event.player.uniqueId)
    }

    @EventHandler
    fun onWorldChange(event: PlayerChangedWorldEvent) {
        // 换世界重置状态（新世界全新开始）
        trackers[event.player.uniqueId]?.let { t ->
            t.bucket = null
            t.recentPlacements = null
            t.lastYaw = null
            t.lastPitch = null
        }
    }
}
